tests/java/verifiers/src/main/java/io/vectorized/reads_writes/Workload.java

```diff
@@ -0,0 +1,437 @@
+    props.put(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, true);
+
+    Producer<String, String> producer = new KafkaProducer<>(props);
+
+    Semaphore rsem;
+    WriterStat stat = null;
+    synchronized (this) {
+      rsem = rsems.get(partition);
+      stat = wstat.get(partition);
+    }
+
+    boolean is_feedback_started = false;
+
+    while (is_active) {
+      if (stat.countWritten >= args.read_write_loop_slack) {
+        if (!is_feedback_started) {
+          is_feedback_started = true;
+          log(partition, "feedback");
+        }
+        rsem.acquire();
+      }
+
+      OpInfo op = new OpInfo();
+      op.id = get_oid();
+      op.started_us = System.nanoTime() / 1000;
+      op.status = OpStatus.WRITING;
+      op.partition = partition;
+      op.offset = -1;
+      synchronized (this) {
+        stat.inflights++;
+        ops.put(op.id, op);
+        oids.get(partition).add(op.id);
+        log(partition, "w\t" + op.id);
+      }
+
+      var callback = new SendCallback(op, stat);
+      ProducerRecord<String, String> record
+          = new ProducerRecord<>(args.topic, partition, "" + op.id, "" + op.id);
+      try {
+        producer.send(record, callback);
+      } catch (Exception e) {
+        callback.onCompletion(null, e);
+      }
+    }
+
+    synchronized (this) {
+      stat.draining = true;
+      if (stat.inflights == 0) {
+        return;
+      }
+    }
+
+    stat.drained.acquire();
+  }
+
+  private void readProcess(int partition) throws Exception {
+    var tp = new TopicPartition(args.topic, partition);
+    var tps = Collections.singletonList(tp);
+
+    Properties props = new Properties();
+    props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, args.brokers);
+    props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, false);
+    props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+    props.put(ConsumerConfig.ISOLATION_LEVEL_CONFIG, "read_committed");
+    props.put(
+        ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG,
+        "org.apache.kafka.common.serialization.StringDeserializer");
+    props.put(
+        ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,
+        "org.apache.kafka.common.serialization.StringDeserializer");
+
+    Queue<Long> oids;
+    Semaphore rsem;
+    WriterStat stat;
+    synchronized (this) {
+      oids = this.oids.get(partition);
+      rsem = rsems.get(partition);
+      stat = wstat.get(partition);
+    }
+
+    KafkaConsumer<String, String> consumer = new KafkaConsumer<>(props);
+    consumer.assign(tps);
+
+    long offset = -1;
+
+    while (true) {
+      if (are_writers_stopped) {
+        synchronized (this) {
+          if (stat.lastOffset <= offset) {
+            break;
+          }
+        }
+      }
+
+      log(partition, "r");
+      ConsumerRecords<String, String> records
+          = consumer.poll(Duration.ofMillis(10000));
+      var it = records.iterator();
+      while (it.hasNext()) {
```

Comment:
for this to be equivalent with existing VerifiableConsumer it would have to commit offsets, tolerate errors (redeliveries) and use consumer groups

Reply:
I agree that RWVerifier doesn't cover consumer groups. But I don't think full equivalence is the right goal, instead of having the verifier we should have a set of specialized verifiers aimed at a particular set of features. With a single verifier we tend not to question how good it is and what it actually covers resulting in a false sense of security.

For example the pair of VerifiableConsumer & VerifiableProducer:

  - ignores violation of order
  - ignores time travel
  - ignores when a consumer reads something what was never written
  - ignores indecisive errors on produce
  - has broken idempotency support (during rebalance a message is allowed to be consumed twice but the validation logic raise an error if there is any duplicate)
  - doesn't validate if a partition is assigned to multiple clients at the same time

RWVerifier isn't affected by all but the last point. And it already found a problem which VerifiableConsumer & VerifiableProducer missed.

Instead of adding weak support of consumer groups to RWVerifier we should add another verifier exclusively targeting the consumer groups unaffected by the existing flaws.

Reply:
What I want to have an example of how to write remove services without relying on piping the output via ssh so that when there is a new PR introducing an end_to_end based test I could point to a RWVerifier and ask the author to do something similar.

What do you think of copying `partition_movement_test` into `partition_movement_rw_test` and marking original depricated until we introduce a verifier with the consumer group support? It sounds ugly but postponing this PR also has cost, we'll continue using the regular approach and we'll depend on it more and more.

Reply:
got it, thank you for explaining all the details i think there is no need to create duplicated tests the one we have now is enough, if the RWVerifier is going to be a goto tool to write EndToEnd tests what do you think about making it a little bit better documented in terms of usage/output and its code cleaned up so that the author of the new features will have an easier job.